python: importBitmaps segmentation fault

* Arch Linux (using my own fontforge-git)

* fontforge 20190413
* build date: 07:51 UTC 30-Jul-2019

Due to [recent changes in pango 1.44](https://gitlab.gnome.org/GNOME/pango/issues/386), it no longer loads PCF or BDF (among others) font formats.  As a result I'm attempting to write a script that can absorb PCF and BDF files and generate an "OpenType Bitmap" font.  I have what appears to work using fontforge's own scripting language but when porting it to the python API I've run into a case of segmentation faults on some PCF files, notably terminus.

```py
#!/usr/bin/env python3

import sys
from fontforge import font


def main(argc, argv):
    if argc == 1:
        print('usage: opentype-bitmap font-size1.pcf [font-size2.pcf ...]',
              file=sys.stderr)
        sys.exit(1)

    n = font()

    for arg in argv[1:]:
        print(f'opentype-bitmap: {arg}: loading bitmap strikes',
              file=sys.stderr)
        n.importBitmaps(arg)

    n.generate(n.fullname + '.otb', 'otb')
    print(n.fullname + '.otb')


if __name__ == '__main__':
    main(len(sys.argv), sys.argv)
```

While this appears to have no issue with many PCF and BDF files, it crashes when attempting to open any one of terminus's PCF files:

```
% file ter-x14n.pcf
ter-x14n.pcf: X11 Portable Compiled Font data, bit: MSB, byte: MSB first

% ./opentype-bitmap.py ter-x14n.pcf
opentype-bitmap: ter-x14n.pcf: loading bitmap strikes
zsh: segmentation fault (core dumped)  ./opentype-bitmap.py ter-x14n.pcf
```

Here is a backtrace:

```
#0  0x00007f53c471826c in FVImportBDF (fv=fv@entry=0x556e62b2e120, filename=filename@entry=0x556e62b2e2a0 "ter-x12n.pcf", ispk=ispk@entry=2, toback=0) at fvimportbdf.c:2306
#1  0x00007f53c47bf2ff in PyFFFont_importBitmaps (self=0x7f53c4b8bef0, args=<optimized out>) at python.c:14140
#2  0x00007f53c5bc85df in _PyMethodDef_RawFastCallKeywords () at /usr/lib/libpython3.7m.so.1.0
#3  0x00007f53c5bfa11f in _PyMethodDescr_FastCallKeywords () at /usr/lib/libpython3.7m.so.1.0
#4  0x00007f53c5bfa2b4 in  () at /usr/lib/libpython3.7m.so.1.0
#5  0x00007f53c5c2ff70 in _PyEval_EvalFrameDefault () at /usr/lib/libpython3.7m.so.1.0
#6  0x00007f53c5be9633 in _PyFunction_FastCallKeywords () at /usr/lib/libpython3.7m.so.1.0
#7  0x00007f53c5bfa1d0 in  () at /usr/lib/libpython3.7m.so.1.0
#8  0x00007f53c5c2fed7 in _PyEval_EvalFrameDefault () at /usr/lib/libpython3.7m.so.1.0
#9  0x00007f53c5be8768 in _PyEval_EvalCodeWithName () at /usr/lib/libpython3.7m.so.1.0
#10 0x00007f53c5be952a in PyEval_EvalCodeEx () at /usr/lib/libpython3.7m.so.1.0
#11 0x00007f53c5c7909c in PyEval_EvalCode () at /usr/lib/libpython3.7m.so.1.0
#12 0x00007f53c5cbd815 in  () at /usr/lib/libpython3.7m.so.1.0
#13 0x00007f53c5cbdd9b in PyRun_FileExFlags () at /usr/lib/libpython3.7m.so.1.0
#14 0x00007f53c5cc48a7 in PyRun_SimpleFileExFlags () at /usr/lib/libpython3.7m.so.1.0
#15 0x00007f53c5cc6aa8 in  () at /usr/lib/libpython3.7m.so.1.0
#16 0x00007f53c5cc6c5c in _Py_UnixMain () at /usr/lib/libpython3.7m.so.1.0
#17 0x00007f53c5e30f83 in __libc_start_main (main=
    0x556e61279020 <main>, argc=11, argv=0x7ffdb0df5ce8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffdb0df5cd8)
    at ../csu/libc-start.c:308
#18 0x0000556e6127905e in _start ()
```

Curiously during building of fontforge I get this error from gcc (there are a few others like it):

```
In function ‘PcfParse’,
    inlined from ‘SFImportBDF’ at fvimportbdf.c:2165:8:
fvimportbdf.c:1808:17: warning: argument 1 range [18446744071562067968, 18446744073709551615] exceeds maximum object size 9223372036854775807 [-Walloc-size-larger-than=]
 1808 |     b->glyphs = calloc(mcnt,sizeof(BDFChar *));
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/glib-2.0/glib/gutils.h:306,
                 from /usr/include/glib-2.0/glib/gthread.h:34,
                 from /usr/include/glib-2.0/glib/gasyncqueue.h:32,
                 from /usr/include/glib-2.0/glib.h:32,
                 from /usr/include/glib-2.0/gobject/gbinding.h:28,
                 from /usr/include/glib-2.0/glib-object.h:23,
                 from /usr/include/glib-2.0/gio/gioenums.h:28,
                 from /usr/include/glib-2.0/gio/giotypes.h:28,
                 from /usr/include/glib-2.0/gio/gio.h:26,
                 from ../inc/ffglib.h:29,
                 from baseviews.h:31,
                 from fvimportbdf.h:4,
                 from fvimportbdf.c:30:
fvimportbdf.c: In function ‘SFImportBDF’:
/usr/include/stdlib.h:542:14: note: in a call to allocation function ‘calloc’ declared here
  542 | extern void *calloc (size_t __nmemb, size_t __size)
      |              ^~~~~~
```

--- a/fontforge/fvimportbdf.c
+++ b/fontforge/fvimportbdf.c
@@ -2297,11 +2297,11 @@ static void SFAddToBackground(SplineFont *sf,BDFFont *bdf);
 
 int FVImportBDF(FontViewBase *fv, char *filename, int ispk, int toback) {
     BDFFont *b, *anyb=NULL;
-    char buf[300];
-    char *eod, *fpt, *file, *full;
+    char *buf, *eod, *fpt, *file, *full, *freeme;
     int fcnt, any = 0;
     int oldenccnt = fv->map->enccount;
 
+    freeme = filename = copy(filename);
     eod = strrchr(filename,'/');
     if (eod != NULL) {
         *eod = '\0';
@@ -2315,16 +2315,18 @@ int FVImportBDF(FontViewBase *fv, char *filename, int ispk, int toback) {
     while (( fpt=strstr(fpt,"; "))!=NULL )
 	{ ++fcnt; fpt += 2; }
 
-    sprintf(buf, _("Loading font from %.100s"), filename);
+    buf = smprintf(_("Loading font from %.100s"), filename);
     ff_progress_start_indicator(10,_("Loading..."),buf,_("Reading Glyphs"),0,fcnt);
     ff_progress_enable_stop(false);
+    free(buf);
 
     do {
 	fpt = strstr(file,"; ");
 	if ( fpt!=NULL ) *fpt = '\0';
 	full = smprintf("%s/%s", filename, file);
-	sprintf(buf, _("Loading font from %.100s"), filename);
+	buf = smprintf(_("Loading font from %.100s"), full);
 	ff_progress_change_line1(buf);
+	free(buf);
 	b = _SFImportBDF(fv->sf,full,ispk,toback, fv->map);
 	free(full);
 	if ( fpt!=NULL ) ff_progress_next_stage();
@@ -2348,6 +2350,7 @@ int FVImportBDF(FontViewBase *fv, char *filename, int ispk, int toback) {
 	ff_post_error( _("No Bitmap Font"), _("Could not find a bitmap font in %s"), filename );
     } else if ( toback )
 	SFAddToBackground(fv->sf,anyb);
+    free(freeme);
 return( any );
 }
 

--- a/fontforge/fvimportbdf.c
+++ b/fontforge/fvimportbdf.c
@@ -2303,9 +2303,15 @@ int FVImportBDF(FontViewBase *fv, char *filename, int ispk, int toback) {
     int oldenccnt = fv->map->enccount;
 
     eod = strrchr(filename,'/');
-    *eod = '\0';
+    if (eod != NULL) {
+        *eod = '\0';
+        file = eod+1;
+    } else {
+        file = filename;
+        filename = ".";
+    }
     fcnt = 1;
-    fpt = eod+1;
+    fpt = file;
     while (( fpt=strstr(fpt,"; "))!=NULL )
 	{ ++fcnt; fpt += 2; }
 
@@ -2313,12 +2319,10 @@ int FVImportBDF(FontViewBase *fv, char *filename, int ispk, int toback) {
     ff_progress_start_indicator(10,_("Loading..."),buf,_("Reading Glyphs"),0,fcnt);
     ff_progress_enable_stop(false);
 
-    file = eod+1;
     do {
 	fpt = strstr(file,"; ");
 	if ( fpt!=NULL ) *fpt = '\0';
-	full = malloc(strlen(filename)+1+strlen(file)+1);
-	strcpy(full,filename); strcat(full,"/"); strcat(full,file);
+	full = smprintf("%s/%s", filename, file);
 	sprintf(buf, _("Loading font from %.100s"), filename);
 	ff_progress_change_line1(buf);
 	b = _SFImportBDF(fv->sf,full,ispk,toback, fv->map);

